app/controllers/add-preprint.js

```diff
@@ -1,87 +1,207 @@
     dropzoneOptions: {
         uploadMultiple: false,
         method: 'PUT'
     },
+
+    _names: ['upload', 'basics', 'subjects', 'authors', 'submit'].map(str => str.capitalize()),
+    valid: new Ember.Object(),
+    init() {
+        this._super(...arguments);
+        if (this.get('session.isAuthenticated')) {
+            this._setCurrentUser();
+        }
+    },
+    _setCurrentUser() {
+        this.get('currentUser').load().then(user => this.set('user', user));
+    },
+    _refreshNodes() {
+        const user = this.get('user');
+        if (user) {
+            loadAll(user, 'nodes', this.get('userNodes'));
+        } else {
+            this.set('userNodes', Ember.A());
+        }
+    },
+    onGetCurrentUser: Ember.observer('user', function() {
+        this._refreshNodes();
+    }),
+
+    filter: [{}, {}, {}],
+    filteredPath: Ember.computed('path', 'filter', 'filter.@each.value', function() {
+        return this.get('path').slice(0, 2).map((path, i) => {
+            if (path.children && this.get(`filter.${i + 1}.value`)) {
+                return {
+                    name: path.name,
+                    children: path.children.filter(child =>
+                        this.get(`filter.${i + 1}.value`).indexOf(child.name || child) !== -1)
+                };
+            }
+            return path;
+        });
+    }),
+    sortedTaxonomies: Ember.computed('taxonomies', 'filter', 'filter.0.value', function() {
+        return [{
+            name: 'a',
+            children: [{
+                name: 'b',
+                children: ['c', 'd', 'e']
+            }, {
+                name: 'f',
+                children: ['g']
+            }],
+        }, {
+            name: 'h',
+            children: [{
+                name: 'i',
+                children: ['j', 'k']
+            }]
+        }, {
+            name: 'l'
+        }].filter(taxonomy =>
+            !this.get('filter.0.value') || taxonomy.name.indexOf(this.get('filter.0.value')) !== -1
+        );
+    }),
+    path: [],
+    selected: new Ember.Object(),
+    sortedSelection: Ember.computed('selected', function() {
+        const sorted = [];
+        const selected = this.get('selected');
+        const flatten = ([obj, name = []]) => {
+            const keys = Object.keys(obj);
+            if (keys.length === 0) {
+                return name.length !== 0 && sorted.pushObject(name);
+            } else {
+                return keys.sort()
+                    .map(key => [obj.get(key), [...name, key]])
+                    .forEach(flatten);
+            }
+        };
+        flatten([selected]);
+        return sorted;
+    }),
+
     actions: {
-        preUpload(comp, drop, file) {
-            this.set('latestFileName', file.name);
-            return new Ember.RSVP.Promise(resolve => {
-                this.set('resolve', resolve);
-            });
+        verify(name, state) {
+            this.get('valid').set(name, state);
```

Comment:
For the future, we'll want to avoid writing our own validation mechanism in favor of a shared standard package. 

(eg no more development along these lines)